### ISSUE TITLE
`no-extra-semi` advice is ill on loops without a body

### ISSUE THREAD

**Issue by silverwind** [eslint/eslint#3075]
```js
for (var i = 0; i <= 5; console.log(i++));
```

[Real life example](https://github.com/nodejs/io.js/blob/c7d8b091623b5dd30977d29f946051c0bf328ad1/lib/cluster.js#L157)

While I agree this is hardly readable and should be refactored, `extra-semi` does incorrectly advice to remove the semicolon here, which would lead to the next line being considered the body of the loop, changing behaviour. The same issue can happen on a `while` loop:
```js
var i;
while (i++ <= 5);
```
[Real life example](https://github.com/nodejs/io.js/blob/c7d8b091623b5dd30977d29f946051c0bf328ad1/lib/url.js#L502)

A `do while` is unaffected because the curly braces are required there. This is on v24.1.

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by michaelficarra** [comment]
The curly braces are not required in a do-while.

**Comment by silverwind** [comment]
Another common pattern that's triggering this rule is calling a function until it returns a falsy value:

```js
while(doSomething());
```

> The curly braces are not required in a do-while.

Please provide an example.

**Comment by michaelficarra** [comment]
`do;while(0)`

**Comment by silverwind** [comment]
Right, valid example there. It should probably be handled as well, even thought it's equivalent to a bodyless `while` there.

**Comment by ilyavolodin** [comment]
Hmm.. This is interesting. While all of the examples above are valid code (although not the most readable), I'm not really sure how we can identify body-less loops in AST. I'm not sure static analysis can pull this off.

**Comment by lo1tuma** [comment]
@ilyavolodin This can be detected by checking if a loop body contains an `EmptyStatement`.

**Comment by ilyavolodin** [comment]
@lo1tuma Ahh.. Nice, didn't realize that. Thanks for pointing that out.

**Comment by lo1tuma** [comment]
Working on this.

**Comment by ilyavolodin** [comment]
It looks like this pattern should be fixed in `curly` rule as well.

**Comment by btmills** [comment]
@ilyavolodin what's the issue with curly? It's working fine for me in both `multi` and `multi-line` modes.

**Comment by ilyavolodin** [comment]
@btmills I opened a new issue for curly here: #3077

**Comment by silverwind** [comment]
Great work, awesome response time - thanks!

I just ran the updated rule against my cases, and they all look to be fixed. :+1:
